### ISSUE TITLE
Support for different set of samples in numeric operations (`add`, `div`, …)

### ISSUE THREAD

**Issue by Luthaf** [lab-cosmo/equistore#288]
Now that we can efficiently compute unions and intersections (#287) for Labels, we can start to lift some restrictions on the operations, in particular about the ability to add/substract/... two `TensorMap` with different labels. 

In these cases, the output should have samples for the union of the input samples, and any missing values should be treated as zeros.

To be done:
- [ ] review all operations and find one which does not support different samples for the inputs
- [ ] update the operations to use union/intersection/... 
    - [ ] add
    - [ ] sub
    - [ ] div
    - [ ] mul
    - [ ] ???

**Comment by PicoCentauri** [comment]
Yes, very good. One question is if operations like `add`/`sub` and friends will always operate on the intersect while for example join should always operate on the `union` or do we also allow it the other way around?

**Comment by agoscinski** [comment]
In join we might give the user the option to choose, see discussion https://github.com/lab-cosmo/equistore/issues/315

**Comment by Luthaf** [comment]
> I wonder if we should not just let it be with an implementation that extends the label entries for TensorMaps with zero values.

We could add this as a new operations (`make_dense`? `fill_samples`?), could you open a separate issue for it? 

I still think there is value in keeping the data in a sparse format as much as possible (in particular for the gradients) and having add/sub/mul/div handle different number of samples in the TensorMap. Join is a different story (it is not a numeric operation), so let's discuss it in a separate issue!
